### PR TITLE
Refactor entity code and add start of Wiki template rendering

### PR DESCRIPTION
**Pull request by oeb25**
This commit is an amalgamation of many changes of varying size (that should have been multiple commits, oops):
- `WikiNodeExt` for convenience methods on `Node`s
- Make `Link` offsets byte-oriented rather than char-oriented
  - _Perhaps there was a reason for char-oriented I missed, and this is inappropriate?_
- Make `Span::add_link` compute the link offsets internally
- Add `Span::add_node` shared by the `From<[Node]>` and `EntityBuilder` methods
- Refactor `EntityIterator` by using the `Deref` impl on `quick_xml::Event`
- Move `EntityBuilder` extraction methods out to take a common `&[Node]` parsed once
  - _These extraction methods have been refactored to use iterators more heavly and wild-cards in matching. Much of this is a style thing, so please let me know and we can revert._
- Add an, early-stage, `render_template` for rendering certain Wiki specific templates such as pronunciations
- Refactor `maybe_prettify_entity_date` and `entity_link_to_html`
  - _Again subjective, so let me know and I'll revert._
- Add expect testing to `EntityBuilder` with a debug Wiki renderer

Please be let me know what changes you like, and which do not fit!

The expect testing was useful for my debugging, but it might be too much of a change to pull in.

### CODE REVIEW

File: core/src/entrypoint/entity.rs [hunk 1]
```diff
@@ -131,203 +124,143 @@ pub struct EntityBuilder {
...
+            Some((key.to_string(), Span::from(value.as_slice())))
+        })
+        .collect();
+    Some(info)
+}
 
-    pub fn append_title(&mut self, text: &str) {
-        self.title.push_str(text);
+fn extract_text(nodes: &[Node]) -> (Option<Span>, Vec<Paragraph>) {
+    fn render_template(name: &str, parameters: &[Parameter]) -> Option<String> {
```
Comment (mikkeldenker) [inline core/src/entrypoint/entity.rs:150]: I would prefer not to have nested functions, even though I see why you have done it here

File: core/src/entrypoint/entity.rs [hunk 2]
```diff
@@ -342,150 +275,225 @@ impl EntityBuilder {
...
+
+    fn ellipsis(s: &str) -> String {
+        if s.len() > 128 {
+            format!("{}…", &s[0..128])
+        } else {
+            s.to_string()
+        }
+    }
+
+    fn check_abstract(title: &str, src: &str, f: impl FnOnce(String)) {
```
Comment (mikkeldenker) [inline core/src/entrypoint/entity.rs:311]: Can we have a more descriptive name than `f` here? It's not obvious without reading about insta what the closure does

**Review by mikkeldenker — CHANGES_REQUESTED** [review]
Overall it looks fine, but there are some minor changes before we can merge.
Its very nice that the name pronunciations are actually displayed now!

File: core/src/entrypoint/entity.rs (hunk 1 above)
Comment (oeb25) [reply core/src/entrypoint/entity.rs:150]: This was left inside due to laziness ;) Moving to module scope.

File: core/src/entrypoint/entity.rs (hunk 2 above)
Comment (oeb25) [reply core/src/entrypoint/entity.rs]: Renamed to `expect` (the name used in rust-analyzer 😉) and added a short doc string.

**Comment by oeb25** [comment]
The link insertion seems to be broken. I will investigate.

Screenshot 2023-08-21 at 19 11 19

**Review by mikkeldenker — APPROVED** [review]

**Pull request merged**